icu4j/main/tests/collate/src/com/ibm/icu/dev/test/util/ULocaleCollationTest.java

```diff
@@ -242,25 +242,25 @@ void checkService(String requestedLocale, ServiceFacade svc,
     public void TestNameList() {
         String[][][] tests = {
                 /* name in French, name in self, minimized, modified */
-                {{"fr-Cyrl-BE", "fr-Cyrl-CA"},
-                    {"Français (cyrillique, Belgique)", "Français (cyrillique, Belgique)", "fr_Cyrl_BE", "fr_Cyrl_BE"},
-                    {"Français (cyrillique, Canada)", "Français (cyrillique, Canada)", "fr_Cyrl_CA", "fr_Cyrl_CA"},
-                },
+//                {{"fr-Cyrl-BE", "fr-Cyrl-CA"},
+//                    {"Français (cyrillique, Belgique)", "Français (cyrillique, Belgique)", "fr_Cyrl_BE", "fr_Cyrl_BE"},
+//                    {"Français (cyrillique, Canada)", "Français (cyrillique, Canada)", "fr_Cyrl_CA", "fr_Cyrl_CA"},
+//                },
                 {{"en", "de", "fr", "zh"},
                     {"Allemand", "Deutsch", "de", "de"},
                     {"Anglais", "English", "en", "en"},
                     {"Chinois", "中文", "zh", "zh"},
                     {"Français", "Français", "fr", "fr"},
                 },
                 // some non-canonical names
-                {{"iw", "iw-US", "no", "no-Cyrl", "in", "in-YU"},
-                    {"Hébreu (États-Unis)", "עברית (ארצות הברית)", "iw_US", "iw_US"},
-                    {"Hébreu (Israël)", "עברית (ישראל)", "iw", "iw_IL"},
-                    {"Indonésien (Indonésie)", "Indonesia (Indonesia)", "in", "in_ID"},
-                    {"Indonésien (Serbie)", "Indonesia (Serbia)", "in_YU", "in_YU"},
-                    {"Norvégien (cyrillique)", "Norsk (kyrillisk)", "no_Cyrl", "no_Cyrl"},
-                    {"Norvégien (latin)", "Norsk (latinsk)", "no", "no_Latn"},
-                },
+//                {{"iw", "iw-US", "no", "no-Cyrl", "in", "in-YU"},
+//                    {"Hébreu (États-Unis)", "עברית (ארצות הברית)", "iw_US", "iw_US"},
+//                    {"Hébreu (Israël)", "עברית (ישראל)", "iw", "iw_IL"},
+//                    {"Indonésien (Indonésie)", "Indonesia (Indonesia)", "in", "in_ID"},
+//                    {"Indonésien (Serbie)", "Indonesia (Serbia)", "in_YU", "in_YU"},
+//                    {"Norvégien (cyrillique)", "Norsk (kyrillisk)", "no_Cyrl", "no_Cyrl"},
+//                    {"Norvégien (latin)", "Norsk (latinsk)", "no", "no_Latn"},
+//                },
```

Comment:
I want to specifically point out my changes to this file, where I simply commented out several tests that were failing.  The tests depend on the old locale-fallback behavior ("just lop components off the end") and don't work with the new default-script logic.  I couldn't figure out enough about how these tests worked to know what a more appropriate fix was, and am hoping somebody familiar with the Java Collator stuff can guide me.

Reply:
Hmmm. You could see where the expected value != Actual value, see if the actual value is what you expect, and modify the test. Or did you try that?

Reply:
This, and the last bullet in the description (“Commented out a few collator tests ... because they're incompatible with the
new fallback logic (specifically, the default-script logic).”) doesn't sound right.

We use different fallback behavior for different “services” / “locale trees”.

For normal locale trees (the main one, curr, lang, zone, ...), we have normal CLDR fallbacks where we do things like skip over a truncation locale that has a different script. For a long time, we have done this via data in the resource bundles using `%%Parent` entries, because we didn't want to hardcode the CLDR fallback behavior. Example: https://github.com/unicode-org/icu/blob/main/icu4c/source/data/locales/zh_Hant.txt gets generated with `%%Parent{"root"}`

As a result, users can write their own resource bundles with their own parent data, and customize ICU data fallbacks by modifying the `%%Parent` entries.

For some locale trees, at least the `coll` tree, we just use truncation fallbacks, because we are not concerned about going to a different script -- the data is used for processing text, not for displaying stuff. Example: https://github.com/unicode-org/icu/blob/main/icu4c/source/data/coll/zh_Hant.txt has no `%%Parent` entry. It has almost no data at all: It only overrides the default collation type (there can be multiple tailorings per language), and *falls back to zh for the actual tailoring data*.

This is specified in the LDML spec: https://www.unicode.org/charts/collation/chart_Ignored.html see “Collation data, however, is an exception. ...”

The CLDR-to-ICU converter clearly has a switch for generating the parent locale data for some trees but not others.

And of course we have supplemental data files that use no fallback at all.

Now, I know that we decided that there are benefits from hardcoding the CLDR parent locale data and fallback behavior, and we have essentially decided that CLDR is now an industry-wide standard, and we don't really know of anyone writing ICU-format resource bundles for their application data. But we need to apply this fallback mechanism only for data trees where this makes sense.

Note that we have an `enum OpenType`, for example in ICUResourceBundle.java. So far, we have used it for other distinctions, such as not using the default locale for `coll`, `brkitr`, and `translit`, see https://www.unicode.org/reports/tr35/tr35.html#411-bundle-vs-item-lookup “The default locale is not even always used ...”

For these trees, we should be loading bundles with OpenType.LOCALE_ROOT.

We could add a new value like LOCALE_ROOT_NO_PARENTS to skip the newly hardcoded CLDR parent locale fallbacks for collation.

Actually, maybe we should treat coll/brkitr/translit the same for both no-default and no-parents. If so, then we could just use LOCALE_ROOT for both. Indeed, https://github.com/unicode-org/icu/blob/main/icu4c/source/data/brkitr/zh_Hant.txt also has no `%%Parent` entry. The translit bundles don't use locale IDs for their names, so the usual CLDR fallback makes no sense there anyway. We should update the LDML spec to align the types of data on no-default = no-parents.

Reply:
>Actually, maybe we should treat coll/brkitr/translit the same for both no-default and no-parents. If so, then we could just use LOCALE_ROOT for both.

I thought about it, and I agree with this.  I changed the ICU4J code to check the open type and bypass the special parent lookup when it's something other than LOCALE_DEFAULT_ROOT (my implementation does still retain the special handling of the script field, as you recommended).  There are two problems with this, though:

1) The unit test failures I commented out still fail, so this change didn't fix that.

This is because that particular test isn't _doing_ parent fallback in the "coll" resource tree, even though it's in the collation tests.  That particular test is testing the _display name_ functionality on Collator.  That's the LocaleDisplayNames stuff, which is in the "locales" resource tree and DOES want the new parent-fallback stuff.  I could find a way to hack the collator stuff to use straight truncation fallback, but I don't think we want to do that; I think we need to fix the tests.  I didn't do that originally because the assumption of how the lookup works is baked so deeply into the design of that test, so I'm hoping you (or somebody else) can give me some advice on the best way to fix this test.  Or I can leave it commented out a filed another ticket.

2) Given that the unit test above doesn't actually exercise the new functionality, I'm not sure anything in our test suite does.  I can (and probably will) update my own new unit test to exercise it a little, but that won't really tell us whether this does what we want in real-world conditions.  Yes, I can add dummy resources over in testData.jar, but that's not really verifying under real-world conditions that it does what we want, either.  Any thoughts on this?

Reply:
To clarify what I'm asking for, `ULocaleCollationTest.TestNameList()` has a list of test cases, each of which has a list of locales, followed by a list of expected values for each of the locales.  The expected-values list has four expected values: name in French, name in self, and the minimized and maximized versions of the locale ID.  In the test cases I've commented out, I believe the problem is the "name in self" value.  If "self" doesn't exist and it uses a non-default script, like `fr_Cyrl_BE`, we fall all the way back to root (or, actually, to the system default locale-- when I run, I actually get "French (Cyrillic, Belgium)").  I guess I could just change the expected values in these tests to look for that, but that feels weird, especially since it'd depend on the default locale.  What do you all think?

Reply:
Thanks for the changes you made.
And yes, I didn't appreciate that these are display name tests, and display names come from regular, display-oriented resource bundles, so the behavior changes make sense.

Reply:
Still not sure how to fix these tests.  The easy fix is just to change the expected results to whatever I'm actually getting (presumably English), but that hard-codes an expectation that the system default locale is English, which seems brittle.  I'm leaving those tests commented out for now, but I can make that change if you think that's the right one to make.

Reply:
Ok. We do test ICU with default locales other than English, so we can't have unguarded tests that only work when it's English.

Reply:
I filed https://unicode-org.atlassian.net/browse/ICU-22152 for somebody to do something smarter with this unit test, so the problem doesn't just fall on the floor.